cosmo_tester/framework/cfy_helper.py

```diff
@@ -97,6 +101,34 @@ def bootstrap(self,
                 task_retry_interval=task_retry_interval,
                 verbose=verbose,
                 debug=debug).wait()
+            self.upload_plugins(cfy)
+
+    def _get_wagons(self, wagons=None):
+        self.logger.info('Downloading Wagons...')
+
+        wagon_paths = []
+        versions_repo_url = 'https://github.com/geokala/' \
+                            'cloudify-versions.git'
+        path = git_helper.clone(
+            versions_repo_url, self._cfy_workdir,
+            branch=os.environ.get('CLOUDIFY_VERSIONS_BRANCH',
+                                  'master'))
+        url_path = os.path.join(path, 'plugin-urls.yaml')
+        plugins = get_yaml_as_dict(url_path)['plugins']
+        for plugin in plugins:
+            if (plugin['name'] or plugin['wgn_url']) in wagons or not wagons:
+                self.logger.info(
+                    'Downloading: {0}...'.format(plugin['wgn_url'])
+                )
+                wagon_paths.append(download_file(plugin['wgn_url']))
+        return wagon_paths
+
+    def upload_plugins(self, cfy, chosen_wagons=None):
+        downloaded_wagon_paths = self._get_wagons(chosen_wagons or [])
+        for wagon in downloaded_wagon_paths:
+            self.logger.info('Uploading {0}'.format(wagon))
+            upload = cfy.plugins.upload(p=wagon, verbose=True)
```

Comment:
It is blocking in one situation: If you run it in an interactive interpreter without assigning its return value, and not in a function.

Otherwise it seems to become non-blocking, which leads to the upload_plugins function returning several minutes before the plugins are actually uploaded.